File: python/2642.design-graph-with-shortest-path-calculator/solution.py
# Created by Jones at 2023/11/11 19:49
# leetgo: dev
# https://leetcode.cn/problems/design-graph-with-shortest-path-calculator/

from typing import *
from leetgo_py import *

from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from copy import deepcopy
from functools import cache, cmp_to_key, lru_cache, reduce
from heapq import heapify, heappop, heappush, heappushpop, heapreplace
from itertools import accumulate, chain, count, pairwise, zip_longest
from math import ceil, comb, floor, gcd, inf, isqrt, log2, perm, sqrt
from operator import xor
from string import ascii_lowercase
from typing import List, Optional

# @lc code=begin
# from sortedcontainers import SortedList


class Graph:
    def __init__(self, n: int, edges: List[List[int]]):
        self.n = n
        self.g = [[] for _ in range(n)]
        for x, y, c in edges:
            self.g[x].append((y, c))

    def addEdge(self, edge: List[int]) -> None:
        x, y, c = edge
        self.g[x].append((y, c))

    def shortestPath(self, node1: int, node2: int) -> int:
        dist = [inf] * self.n
        dist[node1] = 0
        q = [(0, node1)]
        while q:
            d, x = heappop(q)
            for y, c in self.g[x]:
                if d + c < dist[y]:
                    dist[y] = d + c
                    heappush(q, (dist[y], y))

        res = dist[node2]
        if res == inf:
            return -1
        return res


# Your Graph object will be instantiated and called as such:
# obj = Graph(n, edges)
# obj.addEdge(edge)
# param_2 = obj.shortestPath(node1,node2)

# @lc code=end

if __name__ == "__main__":
    ops: List[str] = deserialize("List[str]", read_line())
    params = split_array(read_line())
    output = ["null"]

    constructor_params = split_array(params[0])
    n: int = deserialize("int", constructor_params[0])
    edges: List[List[int]] = deserialize("List[List[int]]", constructor_params[1])
    obj = Graph(n, edges)

    for i in range(1, len(ops)):
        match ops[i]:
            case "addEdge":
                method_params = split_array(params[i])
                edge: List[int] = deserialize("List[int]", method_params[0])
                obj.addEdge(edge)
                output.append("null")
            case "shortestPath":
                method_params = split_array(params[i])
                node1: int = deserialize("int", method_params[0])
                node2: int = deserialize("int", method_params[1])
                ans = serialize(obj.shortestPath(node1, node2))
                output.append(ans)

    print("\noutput:", join_array(output))

File: rust/src/2642.design-graph-with-shortest-path-calculator/solution.rs
// Created by Jones at 2024/03/26 10:59
// leetgo: 1.4.3
// https://leetcode.cn/problems/design-graph-with-shortest-path-calculator/

use anyhow::Result;
use leetgo_rs::*;

// @lc code=begin
use std::cmp::Reverse;
use std::collections::*;
use std::mem::swap;

const INF: i32 = 1e9 as i32;
struct Graph {
    f: Vec<Vec<i32>>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Graph {
    fn new(n: i32, edges: Vec<Vec<i32>>) -> Self {
        let n = n as usize;
        let mut f = vec![vec![INF; n]; n];
        for i in 0..n {
            f[i][i] = 0;
        }

        let mut graph = Self { f };

        for edge in edges {
            graph.add_edge(edge);
        }

        graph
    }

    fn add_edge(&mut self, edge: Vec<i32>) {
        let f = &mut self.f;
        let n = f.len();
        let (x, y, cost) = (edge[0] as usize, edge[1] as usize, edge[2]);
        for i in 0..n {
            for j in 0..n {
                f[i][j] = f[i][j].min(f[i][x] + cost + f[y][j])
            }
        }
    }

    fn shortest_path(&mut self, node1: i32, node2: i32) -> i32 {
        let (x, y) = (node1 as usize, node2 as usize);
        if self.f[x][y] < INF {
            self.f[x][y]
        } else {
            -1
        }
    }
}

// @lc code=end

fn main() -> Result<()> {
    let ops: Vec<String> = deserialize(&read_line()?)?;
    let params = split_array(&read_line()?)?;
    let mut output = Vec::with_capacity(ops.len());
    output.push("null".to_string());

    let constructor_params = split_array(&params[0])?;
    let n: i32 = deserialize(&constructor_params[0])?;
    let edges: Vec<Vec<i32>> = deserialize(&constructor_params[1])?;
    #[allow(unused_mut)]
    let mut obj = Graph::new(n, edges);

    for i in 1..ops.len() {
        match ops[i].as_str() {
            "addEdge" => {
                let method_params = split_array(&params[i])?;
                let edge: Vec<i32> = deserialize(&method_params[0])?;
                obj.add_edge(edge);
                output.push("null".to_string());
            }
            "shortestPath" => {
                let method_params = split_array(&params[i])?;
                let node1: i32 = deserialize(&method_params[0])?;
                let node2: i32 = deserialize(&method_params[1])?;
                let ans: i32 = obj.shortest_path(node1, node2).into();
                output.push(serialize(ans)?);
            }
            _ => panic!("unknown op"),
        }
    }

    println!("\noutput: {}", join_array(output));
    Ok(())
}
